src/Storages/HDFS/StorageHDFSCluster.cpp

```diff
@@ -25,13 +25,42 @@
 #include <Storages/IStorage.h>
 #include <Storages/SelectQueryInfo.h>
 #include <Storages/HDFS/HDFSCommon.h>
+#include <Storages/StorageDictionary.h>
 
 #include <memory>
 
 
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+static ASTPtr addColumnsStructureToQuery(const ASTPtr & query, const String & structure)
```

Comment:
You could take input as `ASTPtr & query` (not const) and modify the query, then we can remove clone & return.